### PR TITLE
feat: implement node inclusion policy in scheduler

### PR DESCRIPTION
**Pull request by kerthcet**
#### What type of PR is this?
/kind feature
/sig scheduling

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/enhancements/issues/3094

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3094
```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 1]
```diff
@@ -1621,6 +1959,31 @@ func TestSingleConstraint(t *testing.T) {
...
+				st.MakePod().Name("p-b1").Node("node-a").Label("foo", "").Obj(),
+				st.MakePod().Name("p-b2").Node("node-b").Label("foo", "").Obj(),
+				st.MakePod().Name("p-y1").Node("node-y").Label("foo", "").Obj(),
+			},
+			enableMinDomains: true,
+			wantStatusCode: map[string]framework.Code{
+				"node-a": framework.Success,
+				"node-b": framework.Success,
+				"node-x": framework.Success,
+				"node-y": framework.Success, // in real case, it's Unschedulable
```
Comment (kerthcet) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1984]: If we can't exclude the filtered out nodes in Filter phase, this test seems like a trick. So I created the pr in https://github.com/kubernetes/kubernetes/pull/108866 cc @alculquicondor

**Comment by kerthcet** [comment]
More unit tests should be covered, otherwise the feature it's completed. (in case someone would like to review this first for close to code frozen)

**Comment by alculquicondor** [comment]
/remove-sig node
The node in the title refers to affinities and taints, which is under SIG scheduling

File: pkg/scheduler/framework/plugins/podtopologyspread/common.go [hunk 2]
```diff
@@ -29,21 +29,33 @@ type topologyPair struct {
 	value string
 }
 
+// nodeInclusionPolicy defines the type of node inclusion policy
+type nodeInclusionPolicy string
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/common.go:33]: why define it again instead of using the core/v1 type?

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 3]
```diff
@@ -587,6 +633,274 @@ func TestPreFilterState(t *testing.T) {
...
+				TpPairToMatchNum: map[topologyPair]int{
+					{key: "node", value: "node-a"}: 1,
+					{key: "node", value: "node-b"}: 2,
+					{key: "node", value: "node-c"}: 0,
+				},
+			},
+			enableNodeInclustionPolicy: true,
+		},
+		{
+			name: "NodeTaintsPolicy ignored",
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:791]: Add a test for NodeTaintsPolicy ignored when the feature is disabled.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1978]: In theory, someone could disable the NodeAffinity plugin and still run Spreading. So we are actually testing a possible scenario and this should be the behavior.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 4]
```diff
@@ -1646,20 +2003,21 @@ func TestSingleConstraint(t *testing.T) {
 
 func TestMultipleConstraints(t *testing.T) {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:2004]: We should have a test for the case where different constraints have different policies.

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go [hunk 5]
```diff
@@ -253,6 +289,208 @@ func TestPreScoreStateEmptyNodes(t *testing.T) {
...
+				TopologyPairToPodCounts: map[topologyPair]*int64{
+					{key: "zone", value: "zone1"}: pointer.Int64Ptr(0),
+					{key: "zone", value: "zone2"}: pointer.Int64Ptr(0),
+				},
+				TopologyNormalizingWeight: []float64{topologyNormalizingWeight(2)},
+			},
+			enableNodeInclustionPolicy: true,
+		},
+		{
+			name: "NodeTaintsPolicy ignored",
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go:462]: also add a case for ignored when the feature is disabled.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 6]
```diff
@@ -260,20 +266,26 @@ func (pl *PodTopologySpread) calPreFilterState(ctx context.Context, pod *v1.Pod)
...
 		// Ensure current node's labels contains all topologyKeys in 'Constraints'.
 		if !nodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
 			return
 		}
 
 		tpCounts := make(map[topologyPair]int, len(constraints))
 		for _, c := range constraints {
+			if c.NodeAffinityPolicy == honor {
+				if match, _ := requiredNodeAffinity.Match(node); !match {
+					continue
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:279]: We should `return`, same for the NodeTaintsPolicy check below, as well as the logic in score.

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring.go [hunk 7]
```diff
@@ -148,14 +154,24 @@ func (pl *PodTopologySpread) PreScore(
...
+				if match, _ := requiredNodeAffinity.Match(node); !match {
+					continue
+				}
+			}
+
+			if c.NodeTaintsPolicy == honor {
+				if _, untolerated := v1helper.FindMatchingUntoleratedTaint(node.Spec.Taints, pod.Spec.Tolerations, nil); untolerated {
+					continue
+				}
+			}
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/scoring.go:173]: It'd be nice to move the logic into a function.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 8]
```diff
@@ -58,23 +65,23 @@ func (p *criticalPaths) sort() {
...
-		want               *preFilterState
+		name                       string
+		pod                        *v1.Pod
+		nodes                      []*v1.Node
+		existingPods               []*v1.Pod
+		objs                       []runtime.Object
+		defaultConstraints         []v1.TopologySpreadConstraint
+		want                       *preFilterState
+		enableMinDomains           bool
+		enableNodeInclustionPolicy bool
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:78]: (not for this line)

Can you add cases to test a pod with multi-constraints (2 is enough), one passed NodeAffinityPolicy/NodeTaintsPolicy, while the other failed. We can add them in `TestPreFilterState()` and `TestMultipleConstraints()`.

**Review by Huang-Wei — COMMENTED** [review]
Thanks @kerthcet .

Could you also add some integration tests?

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go (hunk 1 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1978]: Commented in #108866.

Although it's tricky, it's still possible as Aldo mentioned above. Is it a blocker on your impl.?

Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1978]: No blocker. Let's  cover this scenario mentioned above.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go (hunk 4 above)
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:2004]: yes, still in progress now.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go (hunk 6 above)
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:279]: Why should we directly return. By default, it behaves the same with current implement. When we specified multi constraints, we should still respect others constraints in calculating.

Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:279]: Agree. In the past, the implicit policy applied to the node. Now it is per-constraint.

File: pkg/features/kube_features.go [hunk 9]
```diff
@@ -837,6 +837,14 @@ const (
...
 	// Enable MinDomains in Pod Topology Spread.
 	MinDomainsInPodTopologySpread featuregate.Feature = "MinDomainsInPodTopologySpread"
+
+	// owner: @kerthcet
+	// kep: http://kep.k8s.io/3094
+	// alpha: v1.24
+	//
+	// Allow users to specify whether to take nodeAffinity/nodeTaint into consideration when
+	// calculating pod topology spread skew.
+	PodTopologySpreadNodePolicy featuregate.Feature = "PodTopologySpreadNodePolicy"
```
Comment (denkensk) [inline pkg/features/kube_features.go:847]: I  am curious why you choose a different name style with `MinDomainsInPodTopologySpread`?
Why not use `NodePolicyInPodTopologySpread`?
It is important to have consistent naming of configurations with the same feature.

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring.go (hunk 7 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/podtopologyspread/scoring.go:173]: Bump. This section is duplicated in both score and filter.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go (hunk 6 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:279]: You're right. I was precepted that Node*Policy in multiple constraints always hold the same honor or ignore value.

Can you add UTs to cover this case:

- constraint1: topologyKey=node, NodeAffinityPolicy=Honor
- constarint2: topologyKey=zone, NodeAffinityPolicy=Ignore

Other than this, integration test are still a must.

File: pkg/api/pod/util.go [hunk 10]
```diff
@@ -647,6 +648,53 @@ func dropDisabledCSIVolumeSourceAlphaFields(podSpec, oldPodSpec *api.PodSpec) {
...
+		}
+	}
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PodTopologySpreadNodePolicy) &&
+		!nodeTaintsPolicyInUse(oldPodSpec) &&
+		podSpec != nil {
+		for i := range podSpec.TopologySpreadConstraints {
+			podSpec.TopologySpreadConstraints[i].NodeTaintsPolicy = nil
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/api/pod/util.go:668]: Already modified here https://github.com/kubernetes/kubernetes/pull/108492/files#diff-40853a2fe474b6bde454934dc4e0742a3d9bbf98c31336d8d74520ebe8a2e300R653-R666, different approach but same effect.

[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/api/pod/util.go:668]: I haven't pick the change yet.

File: pkg/features/kube_features.go (hunk 9 above)
Comment (kerthcet) [reply pkg/features/kube_features.go:847]: Good cache, this is former designed in KEP and I was developing this feature the same time with `MinDomainsInPodTopologySpread`, but his feature merged first. I haven't notice the difference here yet. I will follow up this later.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go (hunk 6 above)
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:279]: done :)

**Comment by kerthcet** [comment]
> Thanks @kerthcet .
> 
> Could you also add some integration tests?

All tests completed. Ready for review.

File: test/integration/scheduler/predicates_test.go [hunk 11]
```diff
@@ -1219,25 +1235,251 @@ func TestPodTopologySpreadFilter(t *testing.T) {
...
 			defer testutils.CleanupTest(t, testCtx)
 
-			for i := 0; i < 4; i++ {
-				// Create nodes with labels "zone: zone-{0,1}" and "node: <node name>" to each node.
-				nodeName := fmt.Sprintf("node-%d", i)
-				zone := fmt.Sprintf("zone-%d", i/2)
-				_, err := createNode(cs, st.MakeNode().Name(nodeName).Label("node", nodeName).Label("zone", zone).Obj())
-				if err != nil {
-					t.Fatalf("Cannot create node: %v", err)
+			if len(tt.specificNodes) != 0 {
```
Comment (Huang-Wei) [inline test/integration/scheduler/predicates_test.go:1465]: why introduce this field? I noticed the added cases are all 4 nodes, are labelled the same.

Comment (kerthcet) [reply test/integration/scheduler/predicates_test.go:1465]: Labels are not exactly the same, some nodes hold affinity labels matching selector in `constraint`, what's more, some nodes should have taints to verify `NodeTaintsPolicy`.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 12]
```diff
@@ -530,62 +572,612 @@ func TestPreFilterState(t *testing.T) {
...
+					{key: "zone", value: "zone2"}:  1,
+					{key: "node", value: "node-a"}: 0,
+					{key: "node", value: "node-b"}: 2,
+					{key: "node", value: "node-x"}: 1,
+				},
+			},
+			enableNodeInclustionPolicy: true,
+		},
+		{
+			name: "two Constraints on zone and node, one honored NodeTaintsPolicy, one ignored NodeTaintsPolicy",
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1031]: testcase `one ignored NodeAffinityPolicy, one horned NodeTaintsPolicy` is located at: https://github.com/kubernetes/kubernetes/blob/701d9f3ded7ee94131d4b19a388a97ddaa796768/pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go#L1131-L1181

[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:1031]: Addressed the testcase names

**Comment by alculquicondor** [comment]
Is this rebased on top of the latest API review?

The github review tool is so bad :( I can't even tell which commits I have already seen.

**Comment by kerthcet** [comment]
No, I haven't rebase the latest API review, I just cherry pick commits in that pr. It's the latest.

**Comment by alculquicondor** [comment]
can you rebase?

**Comment by kerthcet** [comment]
> can you rebase?

done

**Review by alculquicondor — COMMENTED** [review]
It looks like you are still missing integration tests for the feature.

**Comment by kerthcet** [comment]
> It looks like you are still missing integration tests for the feature.

I already added here https://github.com/kubernetes/kubernetes/blob/34d2e10d4fbcaa91b0c14e09366f081a2eff110e/test/integration/scheduler/predicates_test.go#L1238-L1281

**Comment by kerthcet** [comment]
kindly ping @alculquicondor as we're close to code freeze deadline. 😀

File: test/integration/scheduler/predicates_test.go [hunk 13]
```diff
@@ -1219,25 +1235,250 @@ func TestPodTopologySpreadFilter(t *testing.T) {
...
+				st.MakeNode().Name("node-1").Label("node", "node-1").Label("zone", "zone-1").Label("foo", "").Obj(),
+				st.MakeNode().Name("node-2").Label("node", "node-2").Label("zone", "zone-1").Label("foo", "").Obj(),
+				st.MakeNode().Name("node-3").Label("node", "node-3").Label("zone", "zone-2").Obj(),
+				st.MakeNode().Name("node-4").Label("node", "node-4").Label("zone", "zone-2").Label("foo", "").Obj(),
+			},
+			candidateNodes:             []string{"node-4"}, // node-3 is filtered out by NodeAffinity plugin
+			enableNodeInclustionPolicy: true,
+		},
+		{
+			name: "NodeAffinityPolicy ignored with nodeAffinity, pods spread across zone as 1/~2~",
```
Comment (alculquicondor) [inline test/integration/scheduler/predicates_test.go:1261]: No need to separate nodeAffinity from node selectors. We run the same code for both.

Integration tests are expensive, so we should try to keep the tests minimal.

File: test/integration/scheduler/predicates_test.go [hunk 14]
```diff
@@ -1219,25 +1235,250 @@ func TestPodTopologySpreadFilter(t *testing.T) {
...
+				st.MakeNode().Name("node-4").Label("node", "node-4").Label("zone", "zone-2").Label("foo", "").Obj(),
+			},
+			candidateNodes:             []string{"node-1", "node-2", "node-4"}, // node-3 is filtered out by TaintToleration plugin
+			enableNodeInclustionPolicy: true,
+		},
+		{
+			// 1. to fulfil "zone" constraint, pods spread across zones as 2/1
+			// 2. to fulfil "node" constraint, pods spread across zones as 1/1/~0~/1
+			// intersection of (1) and (2) returns node-4 as node-3 is filtered out by NodeAffinity plugin.
+			name: "two node inclusion Constraints, zone: honor/ignore, node: honor/ignore",
```
Comment (alculquicondor) [inline test/integration/scheduler/predicates_test.go:1327]: 2 tests cases for "combined policies" should be enough.

File: test/integration/scheduler/predicates_test.go (hunk 11 above)
Comment (alculquicondor) [reply test/integration/scheduler/predicates_test.go:1464]: rename the field to `nodes`

Then, create a variable `defaultNodes` and add it to the test cases that where expecting them.

File: test/integration/scheduler/priorities_test.go [hunk 15]
```diff
@@ -441,7 +441,7 @@ func TestPodTopologySpreadScoring(t *testing.T) {
 		{
 			name: "place pod on a ~0~/1/2/3 cluster with MaxSkew=1, node-1 is the preferred fit",
 			incomingPod: st.MakePod().Namespace(ns).Name("p").Label("foo", "").Container(pause).
-				SpreadConstraint(1, "node", softSpread, st.MakeLabelSelector().Exists("foo").Obj(), nil).
+				SpreadConstraint(1, "node", softSpread, st.MakeLabelSelector().Exists("foo").Obj(), nil, nil, nil).
```
Comment (alculquicondor) [inline test/integration/scheduler/priorities_test.go:444]: Add a couple of scoring tests for the new feature.

**Review by alculquicondor — COMMENTED** [review]
> I already added here

Ah, github reviews not working the way it's supposed to :(

File: test/integration/scheduler/predicates_test.go (hunk 13 above)
Comment (kerthcet) [reply test/integration/scheduler/predicates_test.go:1261]: One test enabled `NodeAffinityPolicy`, the other one disabled.

File: test/integration/scheduler/priorities_test.go (hunk 15 above)
Comment (kerthcet) [reply test/integration/scheduler/priorities_test.go:444]: Sorry, add scoring tests at wrong places. Done :)

**Review by alculquicondor — COMMENTED** [review]
lgtm

@Huang-Wei anything to add before squashing?

**Comment by Huang-Wei** [comment]
LGTM. Please squash.

**Comment by kerthcet** [comment]
squashed

File: pkg/scheduler/framework/plugins/feature/feature.go [hunk 16]
```diff
@@ -20,9 +20,11 @@ package feature
...
-	EnablePodAffinityNamespaceSelector  bool
-	EnablePodDisruptionBudget           bool
-	EnableReadWriteOncePod              bool
-	EnableVolumeCapacityPriority        bool
-	EnableMinDomainsInPodTopologySpread bool
+	EnablePodAffinityNamespaceSelector           bool
+	EnablePodDisruptionBudget                    bool
+	EnableReadWriteOncePod                       bool
+	EnableVolumeCapacityPriority                 bool
+	EnableCSIStorageCapacity                     bool
```
Comment (liggitt) [inline pkg/scheduler/framework/plugins/feature/feature.go:27]: is this a bad rebase? looks unrelated to this feature

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 17]
```diff
@@ -251,29 +256,27 @@ func (pl *PodTopologySpread) calPreFilterState(ctx context.Context, pod *v1.Pod)
...
-		}
+
 		// Ensure current node's labels contains all topologyKeys in 'Constraints'.
 		if !nodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
 			return
 		}
 
 		tpCounts := make(map[topologyPair]int, len(constraints))
 		for _, c := range constraints {
+			if !c.matchHonorNodeInclusionPolicies(pod, node, requiredNodeAffinity) {
```
Comment (liggitt) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:276]: this is being called inside a loop, where before, it was called outside... is that going to significantly change performance when the feature is enabled? what about when the feature is disabled?

Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:276]: Yes, before we verify the nodeAffinity at `node` level, but now NodeInclusionPolicy performs at `constrain` level.
Considering `Constraints`'s amount in real cases will not be quite large, I think it's controllable. Here's the benchmark test results.
1. Master Branch: `Verify nodeAffinity and Skip taints verification`
2. Current Branch: 
   - Feature gate is disabled. `Verify nodeAffinity and Skip taints verification` with new implement. `No significantly performance downgrade` .
   - Feature gate enabled. We didn't run benchmark for this scenario. But the difference locates at the taints verification. Since it looks like there's no parsing involved in taints verification and the taints are limited, `I also think there's no significantly performance downgrade`. (I'll add some tests later to this benchmark to confirm this.)
   

Mater branch:
```
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread
cpu: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
BenchmarkFilter/1000nodes/single-constraint-zone-8                  1458            686904 ns/op
BenchmarkFilter/1000nodes/single-constraint-zone/Clone-8          808203              1693 ns/op
BenchmarkFilter/1000nodes/single-constraint-node-8                  1573            779965 ns/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone-8           12524             95071 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node-8                938           1168812 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-8                12606             95899 ns/op
PASS
ok      k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread     10.881s
```

Current Branch:
```
goos: darwin
goarch: amd64
pkg: k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread
cpu: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
BenchmarkFilter/1000nodes/single-constraint-zone-8                  1682            737574 ns/op
BenchmarkFilter/1000nodes/single-constraint-zone/Clone-8          790586              1589 ns/op
BenchmarkFilter/1000nodes/single-constraint-node-8                  1612            777061 ns/op
BenchmarkFilter/1000nodes/single-constraint-node/Clone-8           12580             94916 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node-8                942           1200677 ns/op
BenchmarkFilter/1000nodes/two-Constraints-zone-node/Clone-8                12678             95820 ns/op
PASS
ok      k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread     12.480s
``` 

We'll also add more benchmark tests when graduating to beta level.

**Comment by kerthcet** [comment]
I think something went wrong with `pull-kubernetes-e2e-gce-100-performance`.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 18]
```diff
@@ -251,29 +256,27 @@ func (pl *PodTopologySpread) calPreFilterState(ctx context.Context, pod *v1.Pod)
...
 	tpCountsByNode := make([]map[topologyPair]int, len(allNodes))
+	requiredNodeAffinity := nodeaffinity.GetRequiredNodeAffinity(pod)
 	processNode := func(i int) {
 		nodeInfo := allNodes[i]
 		node := nodeInfo.Node()
 		if node == nil {
 			klog.ErrorS(nil, "Node not found")
 			return
 		}
-		// In accordance to design, if NodeAffinity or NodeSelector is defined,
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:263]: IIUC, Jordan would like to literally have the same old code running if the feature is disabled (return at this point).

I also considered this initially, but the new code has the same behavior. Since it might not be so trivial to see that it's the same, please just check the feature gate here and return. Similarly, check the feature gate within the loop to skip the checks.

[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:263]: done.

**Comment by kerthcet** [comment]
PTAL @alculquicondor @liggitt

**Review by alculquicondor — COMMENTED** [review]
Thanks,
good to squash

**Comment by kerthcet** [comment]
squashed.

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering.go [hunk 19]
```diff
@@ -251,29 +256,36 @@ func (pl *PodTopologySpread) calPreFilterState(ctx context.Context, pod *v1.Pod)
...
 			return
 		}
 
+		if !pl.enableNodeInclusionPolicyInPodTopologySpread {
+			// spreading is applied to nodes that pass those filters.
+			// Ignore parsing errors for backwards compatibility.
+			if match, _ := requiredNodeAffinity.Match(node); !match {
+				return
+			}
+		}
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:280]: nit: move it above `if !nodeLabelsMatchSpreadConstraints` so to be exactly the same order as before.

(you can update along with the rebase when the API PR gets merged)

Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:280]: I was intentional for fail fast since `nodeLabelsMatchSpreadConstraints(node.Labels, constraints)`  performs better. But to keep the exactly same, I'll update as advised.

**Comment by kerthcet** [comment]
addressed the comments and squashed the commits.

**Comment by kerthcet** [comment]
Test failure fixed by this PR https://github.com/kubernetes/kubernetes/pull/109134

**Comment by kerthcet** [comment]
rebased

**Comment by kerthcet** [comment]
kindly ping @liggitt

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring.go [hunk 20]
```diff
@@ -148,14 +153,16 @@ func (pl *PodTopologySpread) PreScore(
 		if node == nil {
 			return
 		}
-		// (1) `node` should satisfy incoming pod's NodeSelector/NodeAffinity
-		// (2) All topologyKeys need to be present in `node`
-		match, _ := requiredNodeAffinity.Match(node)
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/scoring.go:153]: already addressed.

**Comment by kerthcet** [comment]
cc @alculquicondor @Huang-Wei

File: pkg/scheduler/framework/plugins/podtopologyspread/common.go [hunk 21]
```diff
@@ -33,17 +35,35 @@ type topologyPair struct {
...
-	MinDomains  int32
+	MaxSkew            int32
+	TopologyKey        string
+	Selector           labels.Selector
+	MinDomains         int32
+	NodeAffinityPolicy v1.NodeInclusionPolicy
+	NodeTaintsPolicy   v1.NodeInclusionPolicy
+}
+
+func (tsc *topologySpreadConstraint) matchHonorNodeInclusionPolicies(pod *v1.Pod, node *v1.Node, require nodeaffinity.RequiredNodeAffinity) bool {
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/podtopologyspread/common.go:46]: nit: IMO matchNodeInclusionPolicies reads better, as our logic respects all policies - both Honor and Ignore effects.

**Review by Huang-Wei — COMMENTED** [review]
The implementation lgtm overallly. One nit below.

File: pkg/scheduler/framework/plugins/podtopologyspread/common.go (hunk 21 above)
Comment (kerthcet) [reply pkg/scheduler/framework/plugins/podtopologyspread/common.go:46]: updated, thanks wei.

**Comment by alculquicondor** [comment]
/approve
/lgtm
/hold
for the API change to merge first.

**Comment by kerthcet** [comment]
cc @alculquicondor @liggitt

**Comment by alculquicondor** [comment]
Please rebase.

**Comment by kerthcet** [comment]
still need a lgtm @alculquicondor @Huang-Wei

**Comment by alculquicondor** [comment]
Did you file a bugs for the flaky tests?

**Review by alculquicondor — COMMENTED** [review]
/lgtm

**Comment by kerthcet** [comment]
> Did you file a bugs for the flaky tests?

I tracked the flaky test here https://github.com/kubernetes/kubernetes/issues/109223#issuecomment-1120940566, but I haven't time to dig into the problem.

**Comment by kerthcet** [comment]
Do you hold this PR for special reason @alculquicondor ?

**Comment by alculquicondor** [comment]
/hold cancel
it was just because the API was not merged yet.

Please don't forget to fix the 2 doc comments in the v1 package.
